### PR TITLE
heroku: 7.66.4 -> 8.4.3 (and then 8.7.1)

### PR DESCRIPTION
**Pull request by will**
## Description of changes

Update heroku package. The URL format has changed with the v8 series to include the commit hash (ref: https://github.com/heroku/cli/issues/2477)

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Tested the new heroku command on my arm mac, and a linux nixos vm running on the mac host. Don't have easy access to x86 computers at the moment.

### CODE REVIEW

**Comment by ludat** [comment]
I've tested this on x86_64-linux and it works

File: pkgs/development/tools/heroku/default.nix [hunk 1]
```diff
@@ -1,12 +1,19 @@
...
+  version = "8.4.3";
+  commit = "8226515";
+  hash = "sha256-gbYKA9F3Ue4rcxRhJS4jbIFCEDV2Qy8xEsg0P09eopc=";
+in
+stdenv.mkDerivation {
   pname = "heroku";
-  version = "7.66.4";
+  inherit version;
 
   src = fetchurl {
```
Comment (pbsds) [inline pkgs/development/tools/heroku/default.nix:14]:
```suggestion
  src = fetchzip {
```

I believe this supports `.ex`

Comment (pbsds) [reply pkgs/development/tools/heroku/default.nix:14]: remember to update the hash

**Comment by will** [comment]
I've updated the original commit to use `fetchzip` as suggested. 

I've also pushed a second commit to update to the newer version of `heroku` than was available when I originally made this PR. I figured that leaving it as a second commit rather than changing the version in the original commit was more clear as to what has happened. I wasn't sure if it's better etiquette to open a new PR for a new version bump, or what though.

**Comment by pbsds** [comment]
how about
```diff
diff --git a/pkgs/development/tools/heroku/default.nix b/pkgs/development/tools/heroku/default.nix
index 090e5fa8fd7d..aaf67d4c2406 100644
--- a/pkgs/development/tools/heroku/default.nix
+++ b/pkgs/development/tools/heroku/default.nix
@@ -1,19 +1,12 @@
-{ stdenv, lib, fetchzip, makeWrapper, nodejs }:
+{ stdenv, lib, fetchzip, makeWrapper, nodejs, writeScript }:
 
-let
-  # version and commit pairs can be found in the URLs at
-  # https://cli-assets.heroku.com/versions/heroku-linux-x64-tar-xz.json
-  version = "8.7.1";
-  commit = "3f5e369";
-  hash = "sha256-3pCutQBS8N1Yw4JKTvU046UrOxBi0wLRQywxwezAEeU";
-in
 stdenv.mkDerivation {
   pname = "heroku";
-  inherit version;
+  version = "8.7.1";
 
   src = fetchzip {
-    url = "https://cli-assets.heroku.com/versions/${version}/${commit}/heroku-v${version}-${commit}-linux-x64.tar.xz";
-    inherit hash;
+    url = "https://cli-assets.heroku.com/versions/8.7.1/3f5e369/heroku-v8.7.1-3f5e369-linux-x64.tar.xz";
+    hash = "sha256-3pCutQBS8N1Yw4JKTvU046UrOxBi0wLRQywxwezAEeU";
   };
 
   nativeBuildInputs = [ makeWrapper ];
@@ -29,6 +22,19 @@ stdenv.mkDerivation {
       --set HEROKU_DISABLE_AUTOUPDATE 1
   '';
 
+  passthru.updateScript = writeScript "update-heroku" ''
+    #!/usr/bin/env nix-shell
+    #!nix-shell -I nixpkgs=./. -i bash -p nix-prefetch curl jq common-updater-scripts
+    resp="$(
+        curl -L "https://cli-assets.heroku.com/versions/heroku-linux-x64-tar-xz.json" \
+            | jq '[to_entries[] | { version: .key, url: .value } | select(.version|contains("-")|not)] | sort | .[-1]'
+    )"
+    url="$(jq <<<"$resp" .url --raw-output)"
+    version="$(jq <<<"$resp" .version --raw-output)"
+    hash="$(nix-prefetch fetchzip --url "$(jq <<<"$resp" .url --raw-output)")"
+    update-source-version heroku "$version" "$hash" "$url"
+  '';
+
   meta = {
     homepage = "https://devcenter.heroku.com/articles/heroku-cli";
     description = "Everything you need to get started using Heroku";
```

**Comment by pkoch** [comment]
Since this is a semver-major bump, can we maybe we can split the `updateScript` to its own PR, and get this in? Not trying to be brash, just trying to unblock folks.

**Review by marsam — APPROVED** [review]

**Comment by marsam** [comment]
Sorry for the delay, somehow this slipped under my radar.

Yes, I think the updateScript can be in another PR.

**Comment by pkoch** [comment]
@marsam No need for being sorry! Thank you very much for unblocking this! <3

@will Do you have the patience to spin up another PR, or should I go ahead? I'm getting started with more serious Nix usage, so this might be a good learning experience for me.

**Comment by pbsds** [comment]
Sorry, i forgot to link this: #271632

**Pull request merged**